providers/ldap/ldap_client.go

```diff
@@ -0,0 +1,649 @@
+			attribs = append(attribs, entryAttr)
+		}
+	}
+
+	if !isType(attribs, scope) && !l.hasPermission(attribs) {
+		return nilIdentity, nil
+	}
+
+	filter := "(" + c.ObjectClassAttribute + "=*)"
+	lConn, err := l.newConn()
+	if err != nil {
+		return nilIdentity, fmt.Errorf("Error %v creating connection", err)
+	}
+	defer lConn.Close()
+
+	if strings.EqualFold(c.UserScope, scope) {
+		search = ldap.NewSearchRequest(distinguishedName,
+			ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+			filter,
+			searchConfig.UserSearchAttributes, nil)
+	} else {
+		search = ldap.NewSearchRequest(distinguishedName,
+			ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+			filter,
+			searchConfig.GroupSeachAttributes, nil)
+	}
+
+	result, err := lConn.Search(search)
+	if err != nil {
+		return nilIdentity, fmt.Errorf("Error %v in search query : %v", err, filter)
+	}
+
+	if len(result.Entries) < 1 {
+		return nilIdentity, fmt.Errorf("No identities can be retrieved")
+	} else if len(result.Entries) > 1 {
+		return nilIdentity, fmt.Errorf("More than one result found")
+	}
+
+	entry := result.Entries[0]
+	entryAttributes := entry.Attributes
+	if !l.hasPermission(entry.Attributes) {
+		return nilIdentity, fmt.Errorf("Permission denied")
+	}
+
+	identity, err := l.attributesToIdentity(entryAttributes, distinguishedName, scope)
+	if err != nil {
+		return nilIdentity, err
+	}
+	if identity == nil {
+		return nilIdentity, fmt.Errorf("User Identity not returned for LDAP")
+	}
+	return *identity, nil
+}
+
+func (l *LClient) attributesToIdentity(attribs []*ldap.EntryAttribute, dnStr string, scope string) (*client.Identity, error) {
+	var externalIdType, accountName, externalId, login string
+	user := false
+
+	externalId = dnStr
+	externalIdType = scope
+
+	if isType(attribs, l.Config.UserObjectClass) {
+		for _, attr := range attribs {
+			if attr.Name == l.Config.UserNameField {
+				if len(attr.Values) != 0 {
+					accountName = attr.Values[0]
+				} else {
+					accountName = externalId
+				}
+			}
+			if attr.Name == l.Config.UserLoginField {
+				login = attr.Values[0]
+			}
+		}
+		user = true
+	} else if isType(attribs, l.Config.GroupObjectClass) {
+		for _, attr := range attribs {
+			if attr.Name == l.Config.GroupNameField {
+				if len(attr.Values) != 0 {
+					accountName = attr.Values[0]
+				} else {
+					accountName = externalId
+				}
+			}
+			if attr.Name == l.Config.UserLoginField {
+				if len(attr.Values) > 0 && attr.Values[0] != "" {
+					login = attr.Values[0]
+				}
+			} else {
+				login = accountName
+			}
+		}
+	} else {
+		log.Infof("\nReturning nil from attributesToIdentityFunc\n")
+		return nil, nil
+	}
+
+	identity := &client.Identity{
+		Resource: client.Resource{
+			Type: "identity",
+		},
+		ExternalIdType: externalIdType,
+		ExternalId:     externalId,
+		Name:           accountName,
+		Login:          login,
+		User:           user,
+	}
+	identity.Resource.Id = externalIdType
```

Comment:
"identity.Resource.Id = externalIdType", This seems wrong. It should be externalIDType + ":" + externalID. Check Github/Shibboleth